crates/object-store-cache/src/cache.rs

```diff
@@ -0,0 +1,247 @@
+            },
+        );
+        Arc::new(builder.build()?)
+    }};
+}
+
+/// This was yanked from [object_store::parse_url_opts] with the following changes:
+///
+/// - Build [object_store::ObjectStore] with environment variables
+/// - Return [Arc] instead of [Box]
+#[cfg_attr(
+    not(any(
+        feature = "object-store-aws",
+        feature = "object-store-gcp",
+        feature = "object-store-azure",
+        feature = "object-store-http"
+    )),
+    allow(unused_variables)
+)]
+fn create_object_store<I, K, V>(
+    scheme: ObjectStoreScheme,
+    url: &Url,
+    options: I,
+) -> Result<Arc<DynObjectStore>>
+where
+    I: IntoIterator<Item = (K, V)>,
+    K: AsRef<str>,
+    V: Into<String>,
+{
+    let store: Arc<DynObjectStore> = match scheme {
+        ObjectStoreScheme::Local => Arc::new(LocalFileSystem::new()),
+        #[cfg(feature = "object-store-aws")]
+        ObjectStoreScheme::AmazonS3 => {
+            builder_env_opts!(object_store::aws::AmazonS3Builder, url, options)
+        }
+        #[cfg(feature = "object-store-gcp")]
+        ObjectStoreScheme::GoogleCloudStorage => {
+            builder_env_opts!(object_store::gcp::GoogleCloudStorageBuilder, url, options)
+        }
+        #[cfg(feature = "object-store-azure")]
+        ObjectStoreScheme::MicrosoftAzure => {
+            builder_env_opts!(object_store::azure::MicrosoftAzureBuilder, url, options)
+        }
+        #[cfg(feature = "object-store-http")]
+        ObjectStoreScheme::Http => {
+            let url = &url[..url::Position::BeforePath];
+            let builder = options.into_iter().fold(
+                object_store::http::HttpBuilder::new().with_url(url.to_string()),
+                |builder, (key, value)| match key.as_ref().parse() {
+                    Ok(k) => builder.with_config(k, value),
+                    Err(_) => builder,
+                },
+            );
+            Arc::new(builder.build()?)
+        }
+        s => return Err(Error::ObjectStoreCreate { scheme: s }),
+    };
+    Ok(store)
+}
+
+/// Drop-in replacement for [object_store::parse_url_opts] with caching and env vars.
+///
+/// It will create or retrieve object store based on passed `url` and `options`.
+/// Keeps global cache
+pub async fn parse_url_opts<I, K, V>(
+    url: &Url,
+    options: I,
+) -> crate::Result<(Arc<DynObjectStore>, Path)>
+where
+    I: IntoIterator<Item = (K, V)>,
+    K: AsRef<str>,
+    V: Into<String>,
+{
+    let (scheme, path) = ObjectStoreScheme::parse(url).map_err(object_store::Error::from)?;
+
+    let base_url = url
+        .as_ref()
+        .strip_suffix(path.as_ref())
+        .unwrap_or_default()
+        .try_into()?;
+
+    let object_store_id = ObjectStoreIdentifier::new(base_url, options);
+    let options = object_store_id.get_options();
+
+    {
+        let cache = OBJECT_STORE_CACHE.read().await;
+        if let Some(store) = (*cache).get(&object_store_id) {
+            return Ok((store.clone(), path));
```

Comment:
Can you add some `tracing::debug` statements here and Below when we write to (and sometimes clear) the cache?